spec.bs

```diff
@@ -232,7 +232,183 @@ to conduct an auction to select an advertisement to display to the user, they ca
+  1. If |generateBidOutput|["{{GenerateBidOutput/modelingSignals}}"] [=map/exists=]:
+    1. Let |modelingSignals| be |generateBidOutput|["{{GenerateBidOutput/modelingSignals}}"].
+    1. If |modelingSignals| is greater than 0 and less than 4096:
+      1. Set |bidToSet|'s [=generated bid/modeling signals=] to |modelingSignals|
+  1. Set bid to |bidToSet|.
+  1. Return true.
+</div>
+
+<div algorithm>
+  To <dfn>convert an ad render</dfn> given an {{AdRender}} |adRender|:
+
+  1. If |adRender|["{{AdRender/url}}"] does not [=map/exist=], return false.
+  1. Let |adUrl| be the result of running the [=URL parser=] on |adRender|["{{AdRender/url}}"].
+  1. If |adUrl| is an error, return failure.
+  1. If [=validating an ad url=] given |adUrl| returns false, return failure.
+  1. Let |adDescriptor| be a new [=ad descriptor=].
+  1. Set |adDescriptor|'s [=ad descriptor/url=] to |adUrl|.
+  1. If |adRender|["{{AdRender/width}}"] [=map/exists=]:
+    1. If |adRender|["{{AdRender/height}}"] does not [=map/exist=], return failure.
+    1. Let |width| and |widthUnit| be the result of parsing (TODO) |adRender|["{{AdRender/width}}"].
+    1. Let |height| and |heightUnit| be the result of parsing (TODO)
+      |adRender|["{{AdRender/height}}"].
+    1. Return failure if any of:
+      1. |width| is non-positive, or non-finite.
+      1. |height| is non-positive, or non-finite.
+      1. |widthUnit| is not "px" or "sw".
+      1. |heightUnit| is not "px" or "sw".
+    1. Let |adSize| be a new [=ad size=].
+    1. Set |adSize|'s [=ad size/width=] to |width|, [=ad size/width units=] to |widthUnit|,
+      [=ad size/height=] to |height|, [=ad size/height units=] to |heightUnit|.
+    1. Set |adDescriptor|'s [=ad descriptor/size=] to |adSize|.
+  1. Return |adDescriptor|.
+</div>
+
+<div algorithm>
+  To <dfn>validate an ad url</dfn> given a [=URL=] |adUrl|:
+
+  1. If |adUrl|'s [=url/scheme=] is not "`https`", return false.
+  1. TODO: If it's not in the related interest group's ads, or is excluded, return false.
+  1. Return true.
+</div>
+
+<div algorithm>
+  The <dfn method for="InterestGroupBiddingScriptRunnerGlobalScope">setPriority(|priority|)</dfn>
+  method steps are:
+
+  1. If not first time calling this method:
+    1. Reset priority and return.
+  1. Set priority to |priority|. TODO: figure out how to integrate this into generateBid().
+</div>
+
+<div algorithm>
+  The <dfn method for="InterestGroupBiddingScriptRunnerGlobalScope">setPrioritySignalsOverride(|key|, |priority|)</dfn>
+  method steps are:
+
+  1. Update (insert or assign) prioritySignals[|key|] to |priority|. TODO: figure out how to
+    integrate this into generateBid().
+</div>
+
+<div algorithm>
+  The <dfn method for="InterestGroupReportingScriptRunnerGlobalScope">sendReportTo(|url|)</dfn>
+  method steps are:
+
+  1. [=Throw=] a {{TypeError}} and clears report url if any of the following conditions hold:
+    * The API is called more than once.

```

Comment:
downloaded the diff and tried `git apply` the diff file, but it failed (probably because the local branch is different from what is used for the diff file). So manually applied the suggested changes.